Support upload requests

caalador opened the pull request:
Added framework support for handling upload requests.
Both multipart and XHR2 file uploads are supported.

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/vaadin/flow/2767)
<!-- Reviewable:end -->

denis-anisimov commented:
Reviewed 41 of 42 files at r1.
Review status: 19 of 40 files reviewed at latest revision, 57 unresolved discussions, some commit checks failed.

---

*[flow-documentation/miscellaneous/tutorial-stream-resources.asciidoc, line 2 at r1](https://reviewable.io:443/reviews/vaadin/flow/2767#-KxHytpBm9Pa-HFaF5mH:-KxHytpBm9Pa-HFaF5mI:b-dg02tm) ([raw file](https://github.com/vaadin/flow/blob/273c7357c0ea95cf9430110e76b8ba7e5b6ba5a1/flow-documentation/miscellaneous/tutorial-stream-resources.asciidoc#L2)):*
>  History API

No

---

*[flow-documentation/miscellaneous/tutorial-stream-resources.asciidoc, line 9 at r1](https://reviewable.io:443/reviews/vaadin/flow/2767#-KxHyvHZVVTdZ9CnGb1Z:-KxHyvHZVVTdZ9CnGb1_:b-hlsn6y) ([raw file](https://github.com/vaadin/flow/blob/273c7357c0ea95cf9430110e76b8ba7e5b6ba5a1/flow-documentation/miscellaneous/tutorial-stream-resources.asciidoc#L9)):*
> Using stream resources

We have already a stream resource and most likely a tutorial for it.
Should it says "stream receiver" ?

---

*[flow-server/src/main/java/com/vaadin/server/StreamReceiver.java, line 42 at r1](https://reviewable.io:443/reviews/vaadin/flow/2767#-KxI-B42e8a_Qq_hoZ5L:-KxI-B42e8a_Qq_hoZ5M:bcvcfo) ([raw file](https://github.com/vaadin/flow/blob/273c7357c0ea95cf9430110e76b8ba7e5b6ba5a1/flow-server/src/main/java/com/vaadin/server/StreamReceiver.java#L42)):*
> {@code writer}

What's  the writer ?

---

*[flow-server/src/main/java/com/vaadin/server/StreamRegistration.java, line 46 at r1](https://reviewable.io:443/reviews/vaadin/flow/2767#-KxI-P4EwaQQ-el0ZBmy:-KxI-P4EwaQQ-el0ZBmz:bz918x6) ([raw file](https://github.com/vaadin/flow/blob/273c7357c0ea95cf9430110e76b8ba7e5b6ba5a1/flow-server/src/main/java/com/vaadin/server/StreamRegistration.java#L46)):*
> AbstractStreamResource getResource();

Why not optional ?
May it return `null`?
Missing javadoc doesn't say anything about unregistered resource (as it has been in previous version).

---

*[flow-server/src/main/java/com/vaadin/server/StreamResource.java, line 64 at r1](https://reviewable.io:443/reviews/vaadin/flow/2767#-KxI-hI2veMej-7PiYl6:-KxI-hI2veMej-7PiYl7:b7rj3a7) ([raw file](https://github.com/vaadin/flow/blob/273c7357c0ea95cf9430110e76b8ba7e5b6ba5a1/flow-server/src/main/java/com/vaadin/server/StreamResource.java#L64)):*
> protected

?
The class is `private` .
So it may work only inside this enclosing class but I don't see a new subclass here.

---

*[flow-server/src/main/java/com/vaadin/server/StreamResourceRegistry.java, line 182 at r1](https://reviewable.io:443/reviews/vaadin/flow/2767#-KxI0ErQwVZdb_26wi7m:-KxI0ErQwVZdb_26wi7n:b-7p92et) ([raw file](https://github.com/vaadin/flow/blob/273c7357c0ea95cf9430110e76b8ba7e5b6ba5a1/flow-server/src/main/java/com/vaadin/server/StreamResourceRegistry.java#L182)):*
>  getStreamResource(URI uri) {

Two almost identical methods.

Can the signature be:
```
<T extends AbstractStreamResource> Optional<T> getResource(Class<T>, URI)
```

?

---

*[flow-server/src/main/java/com/vaadin/server/StreamVariable.java, line 31 at r1](https://reviewable.io:443/reviews/vaadin/flow/2767#-KxI0ave7jmJfMoep1D3:-KxI0ave7jmJfMoep1D4:bmj61ka) ([raw file](https://github.com/vaadin/flow/blob/273c7357c0ea95cf9430110e76b8ba7e5b6ba5a1/flow-server/src/main/java/com/vaadin/server/StreamVariable.java#L31)):*
> <p>

empty

---

*[flow-server/src/main/java/com/vaadin/server/StreamVariable.java, line 34 at r1](https://reviewable.io:443/reviews/vaadin/flow/2767#-KxI0cQvDJfP6g5uYTCL:-KxI0cQvDJfP6g5uYTCM:brfbcsd) ([raw file](https://github.com/vaadin/flow/blob/273c7357c0ea95cf9430110e76b8ba7e5b6ba5a1/flow-server/src/main/java/com/vaadin/server/StreamVariable.java#L34)):*
>  6.5

agreed do not use `@since`

---

*[flow-server/src/main/java/com/vaadin/server/StreamVariable.java, line 39 at r1](https://reviewable.io:443/reviews/vaadin/flow/2767#-KxI0hXQaU4JpVjPEj63:-KxI0hXRRGpsbPPX0LBX:b-z7n4yz) ([raw file](https://github.com/vaadin/flow/blob/273c7357c0ea95cf9430110e76b8ba7e5b6ba5a1/flow-server/src/main/java/com/vaadin/server/StreamVariable.java#L39)):*
> by the terminal 

What the terminal is ?

---

*[flow-server/src/main/java/com/vaadin/server/StreamVariable.java, line 65 at r1](https://reviewable.io:443/reviews/vaadin/flow/2767#-KxI0rGjGjgVtobXLfEY:-KxI0rGjGjgVtobXLfEZ:b-g4s1g0) ([raw file](https://github.com/vaadin/flow/blob/273c7357c0ea95cf9430110e76b8ba7e5b6ba5a1/flow-server/src/main/java/com/vaadin/server/StreamVariable.java#L65)):*
> the terminal 

What the terminal is ?

---

*[flow-server/src/main/java/com/vaadin/server/StreamVariable.java, line 81 at r1](https://reviewable.io:443/reviews/vaadin/flow/2767#-KxI0wNV3PJPVXFIFrrx:-KxI0wNV3PJPVXFIFrry:b-2yl0v7) ([raw file](https://github.com/vaadin/flow/blob/273c7357c0ea95cf9430110e76b8ba7e5b6ba5a1/flow-server/src/main/java/com/vaadin/server/StreamVariable.java#L81)):*
> */
>     /**

Double javadocs

---

*[flow-server/src/main/java/com/vaadin/server/StreamVariable.java, line 83 at r1](https://reviewable.io:443/reviews/vaadin/flow/2767#-KxI0zMMxaohR2tXpLjX:-KxI0zMMxaohR2tXpLjY:b-w5whdt) ([raw file](https://github.com/vaadin/flow/blob/273c7357c0ea95cf9430110e76b8ba7e5b6ba5a1/flow-server/src/main/java/com/vaadin/server/StreamVariable.java#L83)):*
> Terminal

terminal

---

*[flow-server/src/main/java/com/vaadin/server/StreamVariable.java, line 147 at r1](https://reviewable.io:443/reviews/vaadin/flow/2767#-KxI18G-tqTWYoksLjD1:-KxI18G-tqTWYoksLjD2:b-vywm10) ([raw file](https://github.com/vaadin/flow/blob/273c7357c0ea95cf9430110e76b8ba7e5b6ba5a1/flow-server/src/main/java/com/vaadin/server/StreamVariable.java#L147)):*
> {@link }

?

---

*[flow-server/src/main/java/com/vaadin/server/communication/StreamReceiverHandler.java, line 82 at r1](https://reviewable.io:443/reviews/vaadin/flow/2767#-KxI1OH0OJZLLmgchCxD:-KxI1OH0OJZLLmgchCxE:b-wyupt9) ([raw file](https://github.com/vaadin/flow/blob/273c7357c0ea95cf9430110e76b8ba7e5b6ba5a1/flow-server/src/main/java/com/vaadin/server/communication/StreamReceiverHandler.java#L82)):*
> .s

?

---

*[flow-server/src/main/java/com/vaadin/server/communication/StreamReceiverHandler.java, line 115 at r1](https://reviewable.io:443/reviews/vaadin/flow/2767#-KxI1YhvXnTCfwfmA6j8:-KxI1YhvXnTCfwfmA6j9:b4lc6qn) ([raw file](https://github.com/vaadin/flow/blob/273c7357c0ea95cf9430110e76b8ba7e5b6ba5a1/flow-server/src/main/java/com/vaadin/server/communication/StreamReceiverHandler.java#L115)):*
> UI.setCurrent(ui);

Shouldn't it be unset in `finally` ? (E.g. prevents memory leaks)

---

*[flow-server/src/main/java/com/vaadin/server/communication/StreamReceiverHandler.java, line 168 at r1](https://reviewable.io:443/reviews/vaadin/flow/2767#-KxI1wZBg7whxU5PtgFO:-KxI1wZBg7whxU5PtgFP:b6pv44f) ([raw file](https://github.com/vaadin/flow/blob/273c7357c0ea95cf9430110e76b8ba7e5b6ba5a1/flow-server/src/main/java/com/vaadin/server/communication/StreamReceiverHandler.java#L168)):*
> try {
>             iter = upload.getItemIterator((HttpServletRequest) request);

Can it be rewritten to use new Java 8 try- resource?

---

*[flow-server/src/main/java/com/vaadin/server/communication/StreamReceiverHandler.java, line 175 at r1](https://reviewable.io:443/reviews/vaadin/flow/2767#-KxI20dLf65JmqXK8tDg:-KxI20dLf65JmqXK8tDh:b-3d93no) ([raw file](https://github.com/vaadin/flow/blob/273c7357c0ea95cf9430110e76b8ba7e5b6ba5a1/flow-server/src/main/java/com/vaadin/server/communication/StreamReceiverHandler.java#L175)):*
>  System.out.println("Form field " + name + " with value "
>                             + Streams.asString(stream) + " detected.");

Debug leftover.

---

*[flow-server/src/main/java/com/vaadin/server/communication/StreamReceiverHandler.java, line 178 at r1](https://reviewable.io:443/reviews/vaadin/flow/2767#-KxI249X4S3S1ygcZfF0:-KxI249X4S3S1ygcZfF1:b-w25xdb) ([raw file](https://github.com/vaadin/flow/blob/273c7357c0ea95cf9430110e76b8ba7e5b6ba5a1/flow-server/src/main/java/com/vaadin/server/communication/StreamReceiverHandler.java#L178)):*
> try {

Try inside try-if.
Please refactor.

---

*[flow-server/src/main/java/com/vaadin/server/communication/StreamReceiverHandler.java, line 187 at r1](https://reviewable.io:443/reviews/vaadin/flow/2767#-KxI1ok6KpyIbOURMhtq:-KxI1ok6KpyIbOURMhtr:b-6cavf7) ([raw file](https://github.com/vaadin/flow/blob/273c7357c0ea95cf9430110e76b8ba7e5b6ba5a1/flow-server/src/main/java/com/vaadin/server/communication/StreamReceiverHandler.java#L187)):*
> System.out.println("File field " + name + " with file name "
>                             + item.getName() + " detected.");

Leftover I guess.

---

*[flow-server/src/main/java/com/vaadin/server/communication/StreamReceiverHandler.java, line 192 at r1](https://reviewable.io:443/reviews/vaadin/flow/2767#-KxI1uujSXv7D38sH38N:-KxI1uujSXv7D38sH38O:b8jfhnl) ([raw file](https://github.com/vaadin/flow/blob/273c7357c0ea95cf9430110e76b8ba7e5b6ba5a1/flow-server/src/main/java/com/vaadin/server/communication/StreamReceiverHandler.java#L192)):*
> e.printStackTrace();

No

---

*[flow-server/src/main/java/com/vaadin/server/communication/StreamReceiverHandler.java, line 297 at r1](https://reviewable.io:443/reviews/vaadin/flow/2767#-KxI2UMY6JphLdcAxjvL:-KxI2UMZaGUK3QGbfA5r:b-d5366q) ([raw file](https://github.com/vaadin/flow/blob/273c7357c0ea95cf9430110e76b8ba7e5b6ba5a1/flow-server/src/main/java/com/vaadin/server/communication/StreamReceiverHandler.java#L297)):*
>  // NOP
I think logging can be used here.
And please don't use `e1`.

---

*[flow-server/src/main/java/com/vaadin/server/communication/StreamReceiverHandler.java, line 337 at r1](https://reviewable.io:443/reviews/vaadin/flow/2767#-KxI2nT9LDWO-40SD0tF:-KxI2nT9LDWO-40SD0tG:bd5zlv5) ([raw file](https://github.com/vaadin/flow/blob/273c7357c0ea95cf9430110e76b8ba7e5b6ba5a1/flow-server/src/main/java/com/vaadin/server/communication/StreamReceiverHandler.java#L337)):*
> session.accessSynchronously(

Is it needed to run this synchronously ?
If it's run from the other thread then this will be blocked until the servlet dispatcher executes this.
If it's the same thread then it doesn't matter .

---

*[flow-server/src/main/java/com/vaadin/server/communication/StreamReceiverHandler.java, line 357 at r1](https://reviewable.io:443/reviews/vaadin/flow/2767#-KxI3RZZ_dktxwU0WQUf:-KxI3RZZ_dktxwU0WQUg:b961uju) ([raw file](https://github.com/vaadin/flow/blob/273c7357c0ea95cf9430110e76b8ba7e5b6ba5a1/flow-server/src/main/java/com/vaadin/server/communication/StreamReceiverHandler.java#L357)):*
> try {

it contains a lot of nested blocks: `if`s, `try`s,  `do`s.

Please refactor it to make shorter and less complex.

---

*[flow-server/src/main/java/com/vaadin/server/communication/StreamRequestHandler.java, line 84 at r1](https://reviewable.io:443/reviews/vaadin/flow/2767#-KxI3rQDE6YGDw9uVN1d:-KxI3rQDE6YGDw9uVN1e:bpnhpkg) ([raw file](https://github.com/vaadin/flow/blob/273c7357c0ea95cf9430110e76b8ba7e5b6ba5a1/flow-server/src/main/java/com/vaadin/server/communication/StreamRequestHandler.java#L84)):*
> StreamReceiver streamReceiver = (StreamReceiver) resource;
<details><summary><i alt="summary">Quoted 9 lines of code…</i></summary>
>                 /*
>                  * URI pattern:
>                  * VAADIN/dynamic/file-upload/[UIID]/[SECKEY]/[NAME]
>                  *
>                  * @see #generateURI
>                  */
>                 // strip away part until the data we are interested starts
>                 int startOfData = pathInfo.indexOf(DYN_RES_PREFIX)
>                         + DYN_RES_PREFIX.length();
</details>
>  
<details><summary><i alt="summary">Quoted 5 lines of code…</i></summary>
>                 String uppUri = pathInfo.substring(startOfData);
>                 // [0] UIid, [1] security key, [2] name
>                 String[] parts = uppUri.split("/", 3);
>                 receiverHandler.handleRequest(session, request, response,
>                         streamReceiver, parts[0], parts[1]);
</details>


Please extract into separate method.

---

*[flow-server/src/main/java/com/vaadin/util/EncodeUtil.java, line 24 at r1](https://reviewable.io:443/reviews/vaadin/flow/2767#-KxI4LIhuprlk4voZKAh:-KxI4LIhuprlk4voZKAi:b-tzd44q) ([raw file](https://github.com/vaadin/flow/blob/273c7357c0ea95cf9430110e76b8ba7e5b6ba5a1/flow-server/src/main/java/com/vaadin/util/EncodeUtil.java#L24)):*
> 7.7.7

version

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/vaadin/flow/2767)*
<!-- Sent from Reviewable.io -->

denis-anisimov commented:
Reviewed 1 of 42 files at r1, 3 of 3 files at r2, 7 of 8 files at r3, 14 of 14 files at r4.
Review status: 38 of 40 files reviewed at latest revision, 56 unresolved discussions, some commit checks failed.

---

*[flow-server/src/main/java/com/vaadin/server/StreamResourceRegistry.java, line 68 at r3](https://reviewable.io:443/reviews/vaadin/flow/2767#-KxI6G-2UAvJMXPL_l6Y:-KxI6G-2UAvJMXPL_l6Z:b-mnw8b5) ([raw file](https://github.com/vaadin/flow/blob/9d5fdbab8e51b8125940231c284aef5a83dcb30e/flow-server/src/main/java/com/vaadin/server/StreamResourceRegistry.java#L68)):*
>  private static final class ReceiverRegistration

There had been a reason why this class was nested  : it has an access to private `resources` from the enclosed class.

I cannot find where this class has been extracted. But how now `unregister` is implemented ?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/vaadin/flow/2767)*
<!-- Sent from Reviewable.io -->

denis-anisimov commented:
Reviewed 1 of 1 files at r5, 1 of 1 files at r6.
Review status: all files reviewed at latest revision, 54 unresolved discussions, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/vaadin/flow/2767)*
<!-- Sent from Reviewable.io -->

caalador commented:
Review status: 35 of 40 files reviewed at latest revision, 58 unresolved discussions.

---

*[flow-documentation/miscellaneous/tutorial-stream-resources.asciidoc, line 2 at r1](https://reviewable.io:443/reviews/vaadin/flow/2767#-KxHytpBm9Pa-HFaF5mH:-KxI66mz5bmQgJijEtKm:b-896fix) ([raw file](https://github.com/vaadin/flow/blob/273c7357c0ea95cf9430110e76b8ba7e5b6ba5a1/flow-documentation/miscellaneous/tutorial-stream-resources.asciidoc#L2)):*
<details><summary><i>Previously, denis-anisimov (Denis) wrote…</i></summary><blockquote>

>  History API

No
</blockquote></details>

Done.

---

*[flow-documentation/miscellaneous/tutorial-stream-resources.asciidoc, line 9 at r1](https://reviewable.io:443/reviews/vaadin/flow/2767#-KxHyvHZVVTdZ9CnGb1Z:-KxI66X4MSMDiv_6HmJJ:b-896fix) ([raw file](https://github.com/vaadin/flow/blob/273c7357c0ea95cf9430110e76b8ba7e5b6ba5a1/flow-documentation/miscellaneous/tutorial-stream-resources.asciidoc#L9)):*
<details><summary><i>Previously, denis-anisimov (Denis) wrote…</i></summary><blockquote>

> Using stream resources

We have already a stream resource and most likely a tutorial for it.
Should it says "stream receiver" ?
</blockquote></details>

Done.

---

*[flow-server/src/main/java/com/vaadin/server/StreamReceiver.java, line 42 at r1](https://reviewable.io:443/reviews/vaadin/flow/2767#-KxI-B42e8a_Qq_hoZ5L:-KxI6YmpXjrHUVXnUBzz:b-896fix) ([raw file](https://github.com/vaadin/flow/blob/273c7357c0ea95cf9430110e76b8ba7e5b6ba5a1/flow-server/src/main/java/com/vaadin/server/StreamReceiver.java#L42)):*
<details><summary><i>Previously, denis-anisimov (Denis) wrote…</i></summary><blockquote>

> {@code writer}

What's  the writer ?
</blockquote></details>

Done.

---

*[flow-server/src/main/java/com/vaadin/server/StreamRegistration.java, line 46 at r1](https://reviewable.io:443/reviews/vaadin/flow/2767#-KxI-P4EwaQQ-el0ZBmy:-KxI6afbPSyVoXI03ezO:b-e9ym2b) ([raw file](https://github.com/vaadin/flow/blob/273c7357c0ea95cf9430110e76b8ba7e5b6ba5a1/flow-server/src/main/java/com/vaadin/server/StreamRegistration.java#L46)):*
<details><summary><i>Previously, denis-anisimov (Denis) wrote…</i></summary><blockquote>

> AbstractStreamResource getResource();

Why not optional ?
May it return `null`?
Missing javadoc doesn't say anything about unregistered resource (as it has been in previous version).
</blockquote></details>

Check newer version. Optional doesn't handle extending classes that well.

---

*[flow-server/src/main/java/com/vaadin/server/StreamResource.java, line 64 at r1](https://reviewable.io:443/reviews/vaadin/flow/2767#-KxI-hI2veMej-7PiYl6:-KxI6qiSIMBgi4UUdmKI:bwf9302) ([raw file](https://github.com/vaadin/flow/blob/273c7357c0ea95cf9430110e76b8ba7e5b6ba5a1/flow-server/src/main/java/com/vaadin/server/StreamResource.java#L64)):*
<details><summary><i>Previously, denis-anisimov (Denis) wrote…</i></summary><blockquote>

> protected

?
The class is `private` .
So it may work only inside this enclosing class but I don't see a new subclass here.
</blockquote></details>

```
 public StreamResource(String name, InputStreamFactory factory) {
        this(name, new Pipe(factory));
        assert name != null;
    }
```

---

*[flow-server/src/main/java/com/vaadin/server/StreamVariable.java, line 31 at r1](https://reviewable.io:443/reviews/vaadin/flow/2767#-KxI0ave7jmJfMoep1D3:-KxI7KXYm_LNZmWLS5zn:b-896fix) ([raw file](https://github.com/vaadin/flow/blob/273c7357c0ea95cf9430110e76b8ba7e5b6ba5a1/flow-server/src/main/java/com/vaadin/server/StreamVariable.java#L31)):*
<details><summary><i>Previously, denis-anisimov (Denis) wrote…</i></summary><blockquote>

> <p>

empty
</blockquote></details>

Done.

---

*[flow-server/src/main/java/com/vaadin/server/StreamVariable.java, line 34 at r1](https://reviewable.io:443/reviews/vaadin/flow/2767#-KxI0cQvDJfP6g5uYTCL:-KxI7J0FzBF8YhunqOO_:b-896fix) ([raw file](https://github.com/vaadin/flow/blob/273c7357c0ea95cf9430110e76b8ba7e5b6ba5a1/flow-server/src/main/java/com/vaadin/server/StreamVariable.java#L34)):*
<details><summary><i>Previously, denis-anisimov (Denis) wrote…</i></summary><blockquote>

>  6.5

agreed do not use `@since`
</blockquote></details>

Done.

---

*[flow-server/src/main/java/com/vaadin/server/StreamVariable.java, line 39 at r1](https://reviewable.io:443/reviews/vaadin/flow/2767#-KxI0hXQaU4JpVjPEj63:-KxI7WR9PKyOiMlyQkC_:bz22rb4) ([raw file](https://github.com/vaadin/flow/blob/273c7357c0ea95cf9430110e76b8ba7e5b6ba5a1/flow-server/src/main/java/com/vaadin/server/StreamVariable.java#L39)):*
<details><summary><i>Previously, denis-anisimov (Denis) wrote…</i></summary><blockquote>

> by the terminal 

What the terminal is ?
</blockquote></details>

See up top:
E.g.
 * in web terminals {@link StreamVariable} can be used to send large files from
 * browsers to the server without consuming large amounts of memory.

---

*[flow-server/src/main/java/com/vaadin/server/StreamVariable.java, line 65 at r1](https://reviewable.io:443/reviews/vaadin/flow/2767#-KxI0rGjGjgVtobXLfEY:-KxI80YfzdmyF5wybi-m:bz22rb4) ([raw file](https://github.com/vaadin/flow/blob/273c7357c0ea95cf9430110e76b8ba7e5b6ba5a1/flow-server/src/main/java/com/vaadin/server/StreamVariable.java#L65)):*
<details><summary><i>Previously, denis-anisimov (Denis) wrote…</i></summary><blockquote>

> the terminal 

What the terminal is ?
</blockquote></details>

See up top:
E.g.
 * in web terminals {@link StreamVariable} can be used to send large files from
 * browsers to the server without consuming large amounts of memory.

---

*[flow-server/src/main/java/com/vaadin/server/StreamVariable.java, line 81 at r1](https://reviewable.io:443/reviews/vaadin/flow/2767#-KxI0wNV3PJPVXFIFrrx:-KxI8K6jmxI-8MxNP6ul:bh2rqef) ([raw file](https://github.com/vaadin/flow/blob/273c7357c0ea95cf9430110e76b8ba7e5b6ba5a1/flow-server/src/main/java/com/vaadin/server/StreamVariable.java#L81)):*
<details><summary><i>Previously, denis-anisimov (Denis) wrote…</i></summary><blockquote>

> */
>     /**

Double javadocs
</blockquote></details>

The first is an internal comment not a javadoc.

---

*[flow-server/src/main/java/com/vaadin/server/StreamVariable.java, line 83 at r1](https://reviewable.io:443/reviews/vaadin/flow/2767#-KxI0zMMxaohR2tXpLjX:-KxI8QSGfHqNyZHq_fJ2:b-896fix) ([raw file](https://github.com/vaadin/flow/blob/273c7357c0ea95cf9430110e76b8ba7e5b6ba5a1/flow-server/src/main/java/com/vaadin/server/StreamVariable.java#L83)):*
<details><summary><i>Previously, denis-anisimov (Denis) wrote…</i></summary><blockquote>

> Terminal

terminal
</blockquote></details>

Done.

---

*[flow-server/src/main/java/com/vaadin/server/StreamVariable.java, line 147 at r1](https://reviewable.io:443/reviews/vaadin/flow/2767#-KxI18G-tqTWYoksLjD1:-KxI8RkaAkeXrNVjGl6C:b-nqysgh) ([raw file](https://github.com/vaadin/flow/blob/273c7357c0ea95cf9430110e76b8ba7e5b6ba5a1/flow-server/src/main/java/com/vaadin/server/StreamVariable.java#L147)):*
<details><summary><i>Previously, denis-anisimov (Denis) wrote…</i></summary><blockquote>

> {@link }

?
</blockquote></details>

newer revision.

---

*[flow-server/src/main/java/com/vaadin/server/communication/StreamReceiverHandler.java, line 82 at r1](https://reviewable.io:443/reviews/vaadin/flow/2767#-KxI1OH0OJZLLmgchCxD:-KxI8_uU-JrMXsT5XZ70:bbg4d05) ([raw file](https://github.com/vaadin/flow/blob/273c7357c0ea95cf9430110e76b8ba7e5b6ba5a1/flow-server/src/main/java/com/vaadin/server/communication/StreamReceiverHandler.java#L82)):*
<details><summary><i>Previously, denis-anisimov (Denis) wrote…</i></summary><blockquote>

> .s

?
</blockquote></details>

miss type.

---

*[flow-server/src/main/java/com/vaadin/server/communication/StreamReceiverHandler.java, line 115 at r1](https://reviewable.io:443/reviews/vaadin/flow/2767#-KxI1YhvXnTCfwfmA6j8:-KxI8qswwKaLpfj5NFfA:b-896fix) ([raw file](https://github.com/vaadin/flow/blob/273c7357c0ea95cf9430110e76b8ba7e5b6ba5a1/flow-server/src/main/java/com/vaadin/server/communication/StreamReceiverHandler.java#L115)):*
<details><summary><i>Previously, denis-anisimov (Denis) wrote…</i></summary><blockquote>

> UI.setCurrent(ui);

Shouldn't it be unset in `finally` ? (E.g. prevents memory leaks)
</blockquote></details>

Done.

---

*[flow-server/src/main/java/com/vaadin/server/communication/StreamReceiverHandler.java, line 175 at r1](https://reviewable.io:443/reviews/vaadin/flow/2767#-KxI20dLf65JmqXK8tDg:-KxI8tUdKe6ZcHMzfJ4F:by9rfkj) ([raw file](https://github.com/vaadin/flow/blob/273c7357c0ea95cf9430110e76b8ba7e5b6ba5a1/flow-server/src/main/java/com/vaadin/server/communication/StreamReceiverHandler.java#L175)):*
<details><summary><i>Previously, denis-anisimov (Denis) wrote…</i></summary><blockquote>

>  System.out.println("Form field " + name + " with value "
>                             + Streams.asString(stream) + " detected.");

Debug leftover.
</blockquote></details>

newer revision

---

*[flow-server/src/main/java/com/vaadin/server/communication/StreamReceiverHandler.java, line 178 at r1](https://reviewable.io:443/reviews/vaadin/flow/2767#-KxI249X4S3S1ygcZfF0:-KxI8vetUGcPgMAd3RGs:boavpvn) ([raw file](https://github.com/vaadin/flow/blob/273c7357c0ea95cf9430110e76b8ba7e5b6ba5a1/flow-server/src/main/java/com/vaadin/server/communication/StreamReceiverHandler.java#L178)):*
> newer revision
newer revision

---

*[flow-server/src/main/java/com/vaadin/server/communication/StreamReceiverHandler.java, line 192 at r1](https://reviewable.io:443/reviews/vaadin/flow/2767#-KxI1uujSXv7D38sH38N:-KxI8xZe899eI7t5IsT6:by9rfkj) ([raw file](https://github.com/vaadin/flow/blob/273c7357c0ea95cf9430110e76b8ba7e5b6ba5a1/flow-server/src/main/java/com/vaadin/server/communication/StreamReceiverHandler.java#L192)):*
<details><summary><i>Previously, denis-anisimov (Denis) wrote…</i></summary><blockquote>

> e.printStackTrace();

No
</blockquote></details>

newer revision

---

*[flow-server/src/main/java/com/vaadin/server/communication/StreamReceiverHandler.java, line 297 at r1](https://reviewable.io:443/reviews/vaadin/flow/2767#-KxI2UMY6JphLdcAxjvL:-KxI8y09C2cbblgXZKN7:by9rfkj) ([raw file](https://github.com/vaadin/flow/blob/273c7357c0ea95cf9430110e76b8ba7e5b6ba5a1/flow-server/src/main/java/com/vaadin/server/communication/StreamReceiverHandler.java#L297)):*
<details><summary><i>Previously, denis-anisimov (Denis) wrote…</i></summary><blockquote>

>  // NOP
I think logging can be used here.
And please don't use `e1`.
</blockquote></details>

newer revision

---

*[flow-server/src/main/java/com/vaadin/server/communication/StreamReceiverHandler.java, line 357 at r1](https://reviewable.io:443/reviews/vaadin/flow/2767#-KxI3RZZ_dktxwU0WQUf:-KxI93kxnLhSp7M9Eofz:by9rfkj) ([raw file](https://github.com/vaadin/flow/blob/273c7357c0ea95cf9430110e76b8ba7e5b6ba5a1/flow-server/src/main/java/com/vaadin/server/communication/StreamReceiverHandler.java#L357)):*
<details><summary><i>Previously, denis-anisimov (Denis) wrote…</i></summary><blockquote>

> try {

it contains a lot of nested blocks: `if`s, `try`s,  `do`s.

Please refactor it to make shorter and less complex.
</blockquote></details>

newer revision

---

*[flow-server/src/main/java/com/vaadin/server/communication/StreamRequestHandler.java, line 84 at r1](https://reviewable.io:443/reviews/vaadin/flow/2767#-KxI3rQDE6YGDw9uVN1d:-KxI9i47a_3Cdd2tDCRX:b-896fix) ([raw file](https://github.com/vaadin/flow/blob/273c7357c0ea95cf9430110e76b8ba7e5b6ba5a1/flow-server/src/main/java/com/vaadin/server/communication/StreamRequestHandler.java#L84)):*
<details><summary><i>Previously, denis-anisimov (Denis) wrote…</i></summary><blockquote>

> StreamReceiver streamReceiver = (StreamReceiver) resource;
>                 /*
>                  * URI pattern:
>                  * VAADIN/dynamic/file-upload/[UIID]/[SECKEY]/[NAME]
>                  *
>                  * @see #generateURI
>                  */
>                 // strip away part until the data we are interested starts
>                 int startOfData = pathInfo.indexOf(DYN_RES_PREFIX)
>                         + DYN_RES_PREFIX.length();
>  
>                 String uppUri = pathInfo.substring(startOfData);
>                 // [0] UIid, [1] security key, [2] name
>                 String[] parts = uppUri.split("/", 3);
>                 receiverHandler.handleRequest(session, request, response,
>                         streamReceiver, parts[0], parts[1]);


Please extract into separate method.
</blockquote></details>

Done.

---

*[flow-server/src/main/java/com/vaadin/util/EncodeUtil.java, line 24 at r1](https://reviewable.io:443/reviews/vaadin/flow/2767#-KxI4LIhuprlk4voZKAh:-KxI9nTzCrL1YlOE2QqP:b-896fix) ([raw file](https://github.com/vaadin/flow/blob/273c7357c0ea95cf9430110e76b8ba7e5b6ba5a1/flow-server/src/main/java/com/vaadin/util/EncodeUtil.java#L24)):*
<details><summary><i>Previously, denis-anisimov (Denis) wrote…</i></summary><blockquote>

> 7.7.7

version
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/vaadin/flow/2767)*
<!-- Sent from Reviewable.io -->

caalador commented:
Review status: 32 of 40 files reviewed at latest revision, 10 unresolved discussions.

---

*[flow-server/src/main/java/com/vaadin/server/StreamResourceRegistry.java, line 182 at r1](https://reviewable.io:443/reviews/vaadin/flow/2767#-KxI0ErQwVZdb_26wi7m:-KxIDybZeWNsWUgRcVHZ:b-896fix) ([raw file](https://github.com/vaadin/flow/blob/273c7357c0ea95cf9430110e76b8ba7e5b6ba5a1/flow-server/src/main/java/com/vaadin/server/StreamResourceRegistry.java#L182)):*
<details><summary><i>Previously, denis-anisimov (Denis) wrote…</i></summary><blockquote>

>  getStreamResource(URI uri) {

Two almost identical methods.

Can the signature be:
```
<T extends AbstractStreamResource> Optional<T> getResource(Class<T>, URI)
```

?
</blockquote></details>

Done.

---

*[flow-server/src/main/java/com/vaadin/server/StreamResourceRegistry.java, line 68 at r3](https://reviewable.io:443/reviews/vaadin/flow/2767#-KxI6G-2UAvJMXPL_l6Y:-KxICWpabNMTmuBEZvei:bhgbfcd) ([raw file](https://github.com/vaadin/flow/blob/9d5fdbab8e51b8125940231c284aef5a83dcb30e/flow-server/src/main/java/com/vaadin/server/StreamResourceRegistry.java#L68)):*
<details><summary><i>Previously, denis-anisimov (Denis) wrote…</i></summary><blockquote>

>  private static final class ReceiverRegistration

There had been a reason why this class was nested  : it has an access to private `resources` from the enclosed class.

I cannot find where this class has been extracted. But how now `unregister` is implemented ?
</blockquote></details>

They have been unified and the getResourceImplementation was faulty

---

*[flow-server/src/main/java/com/vaadin/server/communication/StreamReceiverHandler.java, line 168 at r1](https://reviewable.io:443/reviews/vaadin/flow/2767#-KxI1wZBg7whxU5PtgFO:-KxIF1byGMPDr-6HoijC:bodimqe) ([raw file](https://github.com/vaadin/flow/blob/273c7357c0ea95cf9430110e76b8ba7e5b6ba5a1/flow-server/src/main/java/com/vaadin/server/communication/StreamReceiverHandler.java#L168)):*
<details><summary><i>Previously, denis-anisimov (Denis) wrote…</i></summary><blockquote>

> try {
>             iter = upload.getItemIterator((HttpServletRequest) request);

Can it be rewritten to use new Java 8 try- resource?
</blockquote></details>

sadly FileItemIterator is not autoclosable.

---

*[flow-server/src/main/java/com/vaadin/server/communication/StreamReceiverHandler.java, line 337 at r1](https://reviewable.io:443/reviews/vaadin/flow/2767#-KxI2nT9LDWO-40SD0tF:-KxIF_Hspmc5SFC8Ks8q:b-896fix) ([raw file](https://github.com/vaadin/flow/blob/273c7357c0ea95cf9430110e76b8ba7e5b6ba5a1/flow-server/src/main/java/com/vaadin/server/communication/StreamReceiverHandler.java#L337)):*
<details><summary><i>Previously, denis-anisimov (Denis) wrote…</i></summary><blockquote>

> session.accessSynchronously(

Is it needed to run this synchronously ?
If it's run from the other thread then this will be blocked until the servlet dispatcher executes this.
If it's the same thread then it doesn't matter .
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/vaadin/flow/2767)*
<!-- Sent from Reviewable.io -->

caalador commented:
Review status: 32 of 40 files reviewed at latest revision, 11 unresolved discussions.

---

*[flow-server/src/main/java/com/vaadin/server/communication/StreamReceiverHandler.java, line 187 at r1](https://reviewable.io:443/reviews/vaadin/flow/2767#-KxI1ok6KpyIbOURMhtq:-KxIGL_RiiMPRbGE6g-5:b-896fix) ([raw file](https://github.com/vaadin/flow/blob/273c7357c0ea95cf9430110e76b8ba7e5b6ba5a1/flow-server/src/main/java/com/vaadin/server/communication/StreamReceiverHandler.java#L187)):*
<details><summary><i>Previously, denis-anisimov (Denis) wrote…</i></summary><blockquote>

> System.out.println("File field " + name + " with file name "
>                             + item.getName() + " detected.");

Leftover I guess.
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/vaadin/flow/2767)*
<!-- Sent from Reviewable.io -->

denis-anisimov commented:
Reviewed 4 of 5 files at r7, 6 of 6 files at r8.
Review status: all files reviewed at latest revision, 9 unresolved discussions.

---

*[flow-server/src/main/java/com/vaadin/server/StreamReceiver.java, line 42 at r7](https://reviewable.io:443/reviews/vaadin/flow/2767#-KxIEZYF5IIbD2WllWGk:-KxIEZYGO1af9Iiw2tx-:b-q7bw8s) ([raw file](https://github.com/vaadin/flow/blob/2a11d930226a24383bdc004bf5a85a0269e403f7/flow-server/src/main/java/com/vaadin/server/StreamReceiver.java#L42)):*
> {@code name}

no such parameter

---

*[flow-server/src/main/java/com/vaadin/server/StreamReceiver.java, line 43 at r7](https://reviewable.io:443/reviews/vaadin/flow/2767#-KxIEbpvkVeS2Zcmn_Xb:-KxIEbpvkVeS2Zcmn_Xc:b-lon4a3) ([raw file](https://github.com/vaadin/flow/blob/2a11d930226a24383bdc004bf5a85a0269e403f7/flow-server/src/main/java/com/vaadin/server/StreamReceiver.java#L43)):*
> {@code name}

again

---

*[flow-server/src/main/java/com/vaadin/server/StreamRegistration.java, line 46 at r1](https://reviewable.io:443/reviews/vaadin/flow/2767#-KxI-P4EwaQQ-el0ZBmy:-KxIEmnD9cQXUb61jJmY:b-py294e) ([raw file](https://github.com/vaadin/flow/blob/273c7357c0ea95cf9430110e76b8ba7e5b6ba5a1/flow-server/src/main/java/com/vaadin/server/StreamRegistration.java#L46)):*
<details><summary><i>Previously, caalador wrote…</i></summary><blockquote>

Check newer version. Optional doesn't handle extending classes that well.
</blockquote></details>

Ah, OK.

---

*[flow-server/src/main/java/com/vaadin/server/StreamResource.java, line 64 at r1](https://reviewable.io:443/reviews/vaadin/flow/2767#-KxI-hI2veMej-7PiYl6:-KxIF8GV4DCe6CuhzBBc:b157kup) ([raw file](https://github.com/vaadin/flow/blob/273c7357c0ea95cf9430110e76b8ba7e5b6ba5a1/flow-server/src/main/java/com/vaadin/server/StreamResource.java#L64)):*
<details><summary><i>Previously, caalador wrote…</i></summary><blockquote>

```
 public StreamResource(String name, InputStreamFactory factory) {
        this(name, new Pipe(factory));
        assert name != null;
    }
```
</blockquote></details>

Doesn't answer why the CTOR is protected.
`private` is fine here.

---

*[flow-server/src/main/java/com/vaadin/server/StreamVariable.java, line 31 at r1](https://reviewable.io:443/reviews/vaadin/flow/2767#-KxI0ave7jmJfMoep1D3:-KxIFN8_pJoxQjBLCU8b:bu7xrt2) ([raw file](https://github.com/vaadin/flow/blob/273c7357c0ea95cf9430110e76b8ba7e5b6ba5a1/flow-server/src/main/java/com/vaadin/server/StreamVariable.java#L31)):*
<details><summary><i>Previously, caalador wrote…</i></summary><blockquote>

Done.
</blockquote></details>

The last `<p>` is still empty (it's not needed since it's not closing).

---

*[flow-server/src/main/java/com/vaadin/server/StreamVariable.java, line 39 at r1](https://reviewable.io:443/reviews/vaadin/flow/2767#-KxI0hXQaU4JpVjPEj63:-KxIFXgDRVtP3cS4Rwvs:b-2c0u0p) ([raw file](https://github.com/vaadin/flow/blob/273c7357c0ea95cf9430110e76b8ba7e5b6ba5a1/flow-server/src/main/java/com/vaadin/server/StreamVariable.java#L39)):*
<details><summary><i>Previously, caalador wrote…</i></summary><blockquote>

See up top:
E.g.
 * in web terminals {@link StreamVariable} can be used to send large files from
 * browsers to the server without consuming large amounts of memory.
</blockquote></details>

Hm......
OK, but I would say it's confusing: I have to read the class comment to understand the method comment.

---

*[flow-server/src/main/java/com/vaadin/server/communication/StreamReceiverHandler.java, line 328 at r8](https://reviewable.io:443/reviews/vaadin/flow/2767#-KxIGyKJyqR7-02ejsJe:-KxIGyKJyqR7-02ejsJf:bkofsy4) ([raw file](https://github.com/vaadin/flow/blob/404919ad0718f4b2be2fb7107e6900e95a0a5310/flow-server/src/main/java/com/vaadin/server/communication/StreamReceiverHandler.java#L328)):*
> session.getResourceRegistry().registerResource(streamReceiver)

Actually the code looks weird: it register and immediately unregister the resource.
Does `registerResource` returns the same registration if a receiver has been already registered?

Or I don't understand what's going on here.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/vaadin/flow/2767)*
<!-- Sent from Reviewable.io -->

caalador commented:
Review status: all files reviewed at latest revision, 7 unresolved discussions.

---

*[flow-server/src/main/java/com/vaadin/server/StreamReceiver.java, line 42 at r7](https://reviewable.io:443/reviews/vaadin/flow/2767#-KxIEZYF5IIbD2WllWGk:-KxIHib4zOSjIrIZKlbQ:b-896fix) ([raw file](https://github.com/vaadin/flow/blob/2a11d930226a24383bdc004bf5a85a0269e403f7/flow-server/src/main/java/com/vaadin/server/StreamReceiver.java#L42)):*
<details><summary><i>Previously, denis-anisimov (Denis) wrote…</i></summary><blockquote>

> {@code name}

no such parameter
</blockquote></details>

Done.

---

*[flow-server/src/main/java/com/vaadin/server/StreamReceiver.java, line 43 at r7](https://reviewable.io:443/reviews/vaadin/flow/2767#-KxIEbpvkVeS2Zcmn_Xb:-KxIHivOgxvL3klbVO-T:b-896fix) ([raw file](https://github.com/vaadin/flow/blob/2a11d930226a24383bdc004bf5a85a0269e403f7/flow-server/src/main/java/com/vaadin/server/StreamReceiver.java#L43)):*
<details><summary><i>Previously, denis-anisimov (Denis) wrote…</i></summary><blockquote>

> {@code name}

again
</blockquote></details>

Done.

---

*[flow-server/src/main/java/com/vaadin/server/StreamResource.java, line 64 at r1](https://reviewable.io:443/reviews/vaadin/flow/2767#-KxI-hI2veMej-7PiYl6:-KxIHuynacY94vDYcKjR:b-896fix) ([raw file](https://github.com/vaadin/flow/blob/273c7357c0ea95cf9430110e76b8ba7e5b6ba5a1/flow-server/src/main/java/com/vaadin/server/StreamResource.java#L64)):*
<details><summary><i>Previously, denis-anisimov (Denis) wrote…</i></summary><blockquote>

Doesn't answer why the CTOR is protected.
`private` is fine here.
</blockquote></details>

Done.

---

*[flow-server/src/main/java/com/vaadin/server/StreamVariable.java, line 31 at r1](https://reviewable.io:443/reviews/vaadin/flow/2767#-KxI0ave7jmJfMoep1D3:-KxIHx_VUOWME6Cu4Anb:b-c617iw) ([raw file](https://github.com/vaadin/flow/blob/273c7357c0ea95cf9430110e76b8ba7e5b6ba5a1/flow-server/src/main/java/com/vaadin/server/StreamVariable.java#L31)):*
<details><summary><i>Previously, denis-anisimov (Denis) wrote…</i></summary><blockquote>

The last `<p>` is still empty (it's not needed since it's not closing).
</blockquote></details>

Damn. I mistakenly had done one too many undo:s.

---

*[flow-server/src/main/java/com/vaadin/server/StreamVariable.java, line 39 at r1](https://reviewable.io:443/reviews/vaadin/flow/2767#-KxI0hXQaU4JpVjPEj63:-KxIIWop4DYzEuOnXkkb:b-ndrue3) ([raw file](https://github.com/vaadin/flow/blob/273c7357c0ea95cf9430110e76b8ba7e5b6ba5a1/flow-server/src/main/java/com/vaadin/server/StreamVariable.java#L39)):*
<details><summary><i>Previously, denis-anisimov (Denis) wrote…</i></summary><blockquote>

Hm......
OK, but I would say it's confusing: I have to read the class comment to understand the method comment.
</blockquote></details>

removed all terminal words.

---

*[flow-server/src/main/java/com/vaadin/server/communication/StreamReceiverHandler.java, line 328 at r8](https://reviewable.io:443/reviews/vaadin/flow/2767#-KxIGyKJyqR7-02ejsJe:-KxIKQywWMgT5XH5-ENV:bo51wh8) ([raw file](https://github.com/vaadin/flow/blob/404919ad0718f4b2be2fb7107e6900e95a0a5310/flow-server/src/main/java/com/vaadin/server/communication/StreamReceiverHandler.java#L328)):*
<details><summary><i>Previously, denis-anisimov (Denis) wrote…</i></summary><blockquote>

> session.getResourceRegistry().registerResource(streamReceiver)

Actually the code looks weird: it register and immediately unregister the resource.
Does `registerResource` returns the same registration if a receiver has been already registered?

Or I don't understand what's going on here.
</blockquote></details>

Well the Registration is a wrapper class and as such only matters for the Element that has the Registration.
Updated with a new function.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/vaadin/flow/2767)*
<!-- Sent from Reviewable.io -->

denis-anisimov commented:
Reviewed 5 of 5 files at r9.
Review status: all files reviewed at latest revision, 3 unresolved discussions.

---

*[flow-server/src/main/java/com/vaadin/server/StreamResourceRegistry.java, line 109 at r9](https://reviewable.io:443/reviews/vaadin/flow/2767#-KxIQMOFR4w2uD6RySHh:-KxIQMOFR4w2uD6RySHi:bjdqb2t) ([raw file](https://github.com/vaadin/flow/blob/c01ab6c75dbcc541a9b51202766cbcc0559472d9/flow-server/src/main/java/com/vaadin/server/StreamResourceRegistry.java#L109)):*
> AbstractStreamResource

May be `unregister` method in the `AbstractStreamResource`  instead ?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/vaadin/flow/2767)*
<!-- Sent from Reviewable.io -->

caalador commented:
Review status: all files reviewed at latest revision, 3 unresolved discussions.

---

*[flow-server/src/main/java/com/vaadin/server/StreamResourceRegistry.java, line 109 at r9](https://reviewable.io:443/reviews/vaadin/flow/2767#-KxIQMOFR4w2uD6RySHh:-KxIRKKg5PE21Ebnp-Mz:bcs3gw3) ([raw file](https://github.com/vaadin/flow/blob/c01ab6c75dbcc541a9b51202766cbcc0559472d9/flow-server/src/main/java/com/vaadin/server/StreamResourceRegistry.java#L109)):*
<details><summary><i>Previously, denis-anisimov (Denis) wrote…</i></summary><blockquote>

> AbstractStreamResource

May be `unregister` method in the `AbstractStreamResource`  instead ?
</blockquote></details>

So `AbstractStreamResource` should get a link to the `StreamRegistration`or to this class?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/vaadin/flow/2767)*
<!-- Sent from Reviewable.io -->

denis-anisimov commented:
Review status: all files reviewed at latest revision, 3 unresolved discussions.

---

*[flow-server/src/main/java/com/vaadin/server/StreamResourceRegistry.java, line 109 at r9](https://reviewable.io:443/reviews/vaadin/flow/2767#-KxIQMOFR4w2uD6RySHh:-KxIS9nQuyyYccoGB60m:b-7qovbw) ([raw file](https://github.com/vaadin/flow/blob/c01ab6c75dbcc541a9b51202766cbcc0559472d9/flow-server/src/main/java/com/vaadin/server/StreamResourceRegistry.java#L109)):*
<details><summary><i>Previously, caalador wrote…</i></summary><blockquote>

So `AbstractStreamResource` should get a link to the `StreamRegistration`or to this class?
</blockquote></details>

Well, may be it's not easy from the implementation point of view.

Up to you decide if it's doable.

Not blocking.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/vaadin/flow/2767)*
<!-- Sent from Reviewable.io -->

caalador commented:
Review status: 39 of 40 files reviewed at latest revision, all discussions resolved.

---

*[flow-server/src/main/java/com/vaadin/server/StreamResourceRegistry.java, line 109 at r9](https://reviewable.io:443/reviews/vaadin/flow/2767#-KxIQMOFR4w2uD6RySHh:-KxM-vgbb-wmEfsyu4S7:b4vdt8s) ([raw file](https://github.com/vaadin/flow/blob/c01ab6c75dbcc541a9b51202766cbcc0559472d9/flow-server/src/main/java/com/vaadin/server/StreamResourceRegistry.java#L109)):*
<details><summary><i>Previously, denis-anisimov (Denis) wrote…</i></summary><blockquote>

Well, may be it's not easy from the implementation point of view.

Up to you decide if it's doable.

Not blocking.
</blockquote></details>

I'm going to leave it using the unregister method as this is really internal functionality and only used to disable any uploads to a resource.
Updated it to only accept upload resources.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/vaadin/flow/2767)*
<!-- Sent from Reviewable.io -->

denis-anisimov commented:
Reviewed 1 of 1 files at r10.
Review status: all files reviewed at latest revision, 1 unresolved discussion.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/vaadin/flow/2767)*
<!-- Sent from Reviewable.io -->
